js/content/common.js

```diff
@@ -196,16 +196,45 @@ let ExtensionLayer = (function() {
 
     // NOTE: use cautiously!
     // Run script in the context of the current tab
-    self.runInPageContext = function(fun){
-        let script  = document.createElement('script');
-        script.textContent = '(' + fun + ')();';
+    self.runInPageContext = function(fun) {
+        let script  = document.createElement("script");
+        script.textContent = '(' + fun + ")();";
         document.documentElement.appendChild(script);
         script.parentNode.removeChild(script);
     };
 
     return self;
 })();
 
+class Messenger {
+    static sendMessage(msgID, info) {
+        window.postMessage({
+            type: "es_" + msgID,
+            information: info
+        }, '*');
+    }
+
+    static addMessageListener(msgID, fn, autoRemove) {
+        let callback = function(e) {
+            if (e.source !== window) { return; }
+            if (!e.data.type) { return; }
+            if (e.data.type === "es_" + msgID) {
+                fn(e.data.information);
+                if (autoRemove) {
+                    window.removeEventListener("message", callback);
+                }
+            }
+        }
+        window.addEventListener("message", callback);
+    }
+}
```

Comment:
Hmm. Do we do this:
```
Messenger.sendMessage("message", data);
Messenger.addEventListener("message", callback, true);
```
often enough that it makes sense for a response callback parameter on `sendMessage()`? ie:
```
static sendMessage(msgID, info, fn) {
  if (fn) {
     Messenger.addMessageListener(msgID, fn, true);
  }
  ...
}
```

Reply:
I think we're only doing this once with the user notes, where the page context sends a message before the modal is trying to get dismissed.
Then the extension saves the note and replies with an "OK" signal, so the modal can be closed now.
So I don't think it's worth it